README.md
# flutter_tags
[![pub package](https://img.shields.io/badge/pub-0.4.9+1-orange.svg)](https://pub.dartlang.org/packages/flutter_tags)
[![Awesome Flutter](https://img.shields.io/badge/Awesome-Flutter-blue.svg?longCache=true&style=flat-square)](https://github.com/Solido/awesome-flutter#ui)
[![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/dnag88)

Create beautiful tags quickly and easily.

## Installing
Add this to your package's pubspec.yaml file:
```dart
dependencies:
  flutter_tags: "^0.4.9+1"
```


## DEMO

<div align="center">
<table>
<tr>
<td style="text-align:center">
 <img width = "250px" src="https://github.com/Dn-a/flutter_tags/raw/master/repo-file/img/example0.4.0_1.gif" />
 </td>
 <td style="text-align:center">
  <img width = "250px" src="https://github.com/Dn-a/flutter_tags/raw/master/repo-file/img/example0.4.0_2.gif" />
  </td>
</tr>
</table>
</div>


## Simple usage
```dart

import 'package:flutter_tags/flutter_tags.dart';
.
.
.
List _items;
double _fontSize = 14;

@override
void initState(){
    super.initState();
    // if you store data on a local database (sqflite), then you could do something like this
    Model().getItems().then((items){
            _items = items;
        });
}

@override
Widget build(BuildContext context) {
    return Tags(
      key:_tagStateKey,
      textField: TagsTextField(
        textStyle: TextStyle(fontSize: _fontSize),
        constraintSuggestion: true, suggestions: [],
        //width: double.infinity, padding: EdgeInsets.symmetric(horizontal: 10),
        onSubmitted: (String str) {
          // Add item to the data source.
          setState(() {
              // required
            _items.add(str);
          });
        },
      ),
      itemCount: _items.length, // required
      itemBuilder: (int index){          
            final item = _items[index];
    
            return ItemTags(
                  // Each ItemTags must contain a Key. Keys allow Flutter to
                  // uniquely identify widgets.
                  key: Key(index.toString()),
                  index: index, // required
                  title: item.title,
                  active: item.active,
                  customData: item.customData,
                  textStyle: TextStyle( fontSize: _fontSize, ),
                  combine: ItemTagsCombine.withTextBefore,
                  image: ItemTagsImage(
                    image: AssetImage("img.jpg") // OR NetworkImage("https://...image.png")
                  ), // OR null,
                  icon: ItemTagsIcon(
                    icon: Icons.add,
                  ), // OR null,
                  removeButton: ItemTagsRemoveButton(
                    onRemoved: (){
                        // Remove the item from the data source.
                        setState(() {
                            // required
                            _items.removeAt(index);
                        });
                        //required
                        return true;
                    },
                  ), // OR null,
                  onPressed: (item) => print(item),
                  onLongPressed: (item) => print(item),
            );
    
      },
    );    
}

final GlobalKey<TagsState> _tagStateKey = GlobalKey<TagsState>();
// Allows you to get a list of all the ItemTags
_getAllItem(){
    List<Item> lst = _tagStateKey.currentState?.getAllItem;
    if(lst!=null)
        lst.where((a) => a.active==true).forEach( ( a) => print(a.title));        
}
```
## Wrapped widget example
You are free to wrap ItemTags () inside another widget
```dart
Tags(  
      itemCount: items.length, 
      itemBuilder: (int index){ 
          return Tooltip(
          message: item.title,
          child:ItemTags(
            title:item.title,
          )
          );
      },
    );    
```
### Tags() parameters
|PropName|Description|default value|
|:-------|:----------|:------------|
|`columns`|*Possibility to set number of columns when necessary*|null|
|`itemCount`|*Tag number to display*|required|
|`symmetry`|*Ability to view and scroll tags horizontally*|false|
|`horizontalScroll`|*Offset drawer width*|0.4|
|`heightHorizontalScroll`|*height for HorizontalScroll to set to display tags correctly*|60|
|`spacing`|*Horizontal space between the tags*|6|
|`runSpacing`|*Vertical space between the tags*|14|
|`alignment`|*Horizontal WrapAlignment*|WrapAlignment.center|
|`runAlignment`|*Vertical WrapAlignment*|WrapAlignment.center|
|`direction`|*Direction of the ItemTags*|Axis.horizontal|
|`verticalDirection`|*Iterate Item from the lower to the upper direction or vice versa*|VerticalDirection.down|
|`textDirection`|*Text direction of  the ItemTags*|TextDirection.ltr|
|`itemBuilder`|*tag generator*||
|`textField`|*add textField*|TagsTextFiled()|
|`containerDecoration`|*add tags container decoration*|null|
|`containerPadding`|*add tags container padding*|null|

### ItemTags() parameters
* `index` - *required*
* `title` - *required*
* `textScaleFactor` - *custom textScaleFactor*
* `active` - *bool value (default true)*
* `pressEnabled` - *active onPress tag ( default true)*
* `customData` - *Possibility to add any custom value in customData field, you can retrieve this later. A good example: store an id from Firestore document.*
* `textStyle` - *textStyle()*
* `alignment` - *MainAxisAlignment ( default MainAxisAlignment.center)*
* `combine` - * ability to combine text, icons, images in different ways ( default  ItemTagsCombine.imageOrIconOrText)*
* `icon` - *ItemTagsIcon()*
* `image` - *ItemTagsImage()*
* `removeButton` - *ItemTagsRemoveButton()*
* `borderRadius` - *BorderRadius*
* `border` - *custom border-side*
* `padding` - *default EdgeInsets.symmetric(horizontal: 7, vertical: 5)*
* `elevation` - *default 5*
* `singleItem` - *default false*
* `textOverflow` - *default TextOverflow.fade*
* `textColor` - *default Colors.black*
* `textActiveColor` - *default  Colors.white*
* `color` - *default Colors.white*
* `activeColor` - *default Colors.blueGrey*
* `highlightColor` - 
* `splashColor` - 
* `colorShowDuplicate` - *default  Colors.red*
* `onPressed` - *callback*
* `onLongPressed` - *callback*
* `onRemoved` - *callback*


## Donate
It takes time to carry on this project. If you found it useful or learned something from the source code please consider the idea of donating 5, 20, 50 € or whatever you can to support the project.
- [![Donate](https://img.shields.io/badge/Donate-PayPal-green.svg)](https://www.paypal.me/dnag88)

## Issues
If you encounter problems, open an issue. Pull request are also welcome.


CHANGELOG.md
## [0.4.9+2] - 2021-03-25.
* `[Updated]` - Add tags container decoration.
* `[Updated]` - Add tags container padding.

## [0.4.9+1] - 2020-11-11.

* `[Fixed]` - Issue #66.
* `[Fixed]` - `DataList` null id.

## [0.4.9] - 2020-09-15.

* `[Added]` - Possibility to to define the padding of the `textField`.

## [0.4.8+2] - 2020-06-02.

* `[Updated]` - Documentation.

## [0.4.8+1] - 2020-05-13.

* `[Updated]` - Documentation.

## [0.4.8] - 2020-03-07.

* `[Updated]` - Documentation.

## [0.4.7] - 2020-02-26.

* `[Updated]` -  Documentation.
* General improvement of the code.

## [0.4.6] - 2020-02-26.

* `[Fixed]` -  some problem.
* `[Added]` - Possibility to disabled/enabled `textField` field - Issue #36.
* `[Added]` - Possibility to insert tags not present in the list of suggestions with the `constraintSuggestion` field - Issue #33.
* `onRemoved` field has been moved inside `ItemTagsRemoveButton()` to maintain consistency.
* `onRemoved` now has a bool return type (ex: when you want to add a control before removing a tag).

## [0.4.5] - 2019-10-30.

* Renamed `TagstextField` to `TagsTextField`.
* Added `textCapitalization` field in `TagsTextField`.

## [0.4.4] - 2019-10-16.

* New Features and Documentation Updates.

* Added `child` field in `ItemTagsImage()`, allows more control over the images.
* Fixed alignment `textField` and `suggestions` - Issue #31.
* Added the possibility to get a list of all the `ItemTags` via GlobalKey<TagsState>.

## [0.4.3] - 2019-07-28.

* Minor Update Documentation.
* Fixed some problem.
* removed the 'position' parameter  in TagsTexField () because it is not essential. the same result can be obtained by setting the verticalDirection and TextDirection parameters in Tags ().

## [0.4.2] - 2019-07-27.

* Update Documentation.
* General improvement of the code.

## [0.4.1] - 2019-07-27.

* Minor Update Documentation.

## [0.4.0] - 2019-07-27.

* **Improvements, new Structure, new Features and Documentation Updates**
*
* `SelectableTags` and `InputTags`have been removed, now there is only one widget. **Tags()**
* Now it is possible to personalize every single tag, with the possibility of adding icons, images and a removal button.
* Possibility to display the list with horizontal scroll.

## [0.3.2] - 2019-06-20.

* Issue #14 and #13 fixed.
* Added `customData` field in Tag Class.

## [0.3.1] - 2019-05-13.

* General improvement of the code.

## [0.3.0] - 2019-04-09.

* **New features**
* Possibility to create a customizable popupMenu.
* **SelectableTags**. Possibility to set a color and an activeColor for each tag.
* **InputTags**. Possibility to hide textField.

## [0.2.4] - 2019-04-08.

* fixed some problem.

## [0.2.3] - 2019-04-05.

* General improvement of the code.
* OnInsert, onDelete and onPressed are now optional.

## [0.2.2] - 2019-03-02.

* Added property `textStyle` in InputTags. NOTE: `textColor` has been removed. now it can be set with `textStyle`.
* added property `textStyle` in SelectableTags. NOTE: if you set `color` in it will be ignored, you must use `textColor` `textActiveColor`.
* Created InputSuggestions. Return suggestions in the TextField. Is not complete, soon the list of suggestions will be implemented.
* General improvement of the code.

## [0.2.1] - 2019-03-01.

* The code has been largely rewritten.
* Now the Tag width calculation is very accurate.

## [0.2.0] - 2019-02-24.

* Improved tag width calculation; 
* Possibility to change the margin and padding of the close icon ( InputTags ).

## [0.1.9] - 2019-02-23.

* Width calculated based on the byte length of the title; 
* When the orientation changes, a recalculation of the screen width is performed.

## [0.1.8] - 2019-02-16.

* Improvement of library documentation.

## [0.1.7] - 2019-02-07.

* Added new feature SingleItem on SelectableTags; 
* Possibility to change color/background-color icon on InputTags; - General improvement of the code.

## [0.1.6] - 2019-01-07.

* Fixed error "Infinity or NaN toInt" on InputTags; - general improvement of the code.

## [0.1.5] - 2018-12-24.

* General improvement of the code.

## [0.1.4] - 2018-12-18.

* Added new features.

## [0.1.3] - 2018-12-16.

* Added new highlight feature (InputTags) - general improvement of the code.

## [0.1.2] - 2018-12-15.

* Add InputTags Widget - Improved documentation.

## [0.1.1] - 2018-12-08.

* Improved documentation.

## [0.1.0] - 2018-12-08.

* Did some changing readme.

## [0.0.1] - 2018-12-08.

* Created Selectable Tags.


lib/input_tags.dart
import 'package:flutter/material.dart';
import 'package:flutter_tags/src/suggestions.dart';
import 'package:flutter_tags/src/text_util.dart';

/// Callback
typedef void OnDelete(String tags);
typedef void OnInsert(String tags);

/// PopupMenuBuilder
typedef List<PopupMenuEntry> PopupMenuBuilder(String tag);
typedef void PopupMenuOnSelected(int id, String tag);

class InputTags extends StatefulWidget {
  InputTags(
      {@required this.tags,
      this.columns = 4,
      this.autofocus,
      this.inputDecoration,
      this.maxLength,
      this.keyboardType,
      this.height,
      this.borderRadius,
      this.boxShadow,
      this.placeholder,
      this.symmetry = false,
      this.textFieldHidden = false,
      this.margin,
      this.padding,
      this.alignment,
      this.offset,
      this.duplicate = false,
      this.fontSize = 14,
      this.textOverflow,
      this.lowerCase = false,
      this.textStyle,
      this.autocorrect,
      this.icon,
      this.iconSize,
      this.iconPadding,
      this.iconMargin,
      this.iconColor,
      this.iconBackground,
      this.iconBorderRadius,
      this.color,
      this.backgroundContainer,
      this.highlightColor,
      this.suggestionsList,
      this.onDelete,
      this.onInsert,
      this.popupMenuBuilder,
      this.popupMenuOnSelected,
      Key key})
      : assert(tags != null),
        super(key: key);

  ///List of [Tag] object
  final List<String> tags;

  ///specific number of columns
  final int columns;

  ///autofocus InputField
  final bool autofocus;

  ///decoration InputField
  final InputDecoration inputDecoration;

  ///max-length InputField
  final int maxLength;

  ///keyboard InputField
  final TextInputType keyboardType;

  ///customize the height of the tag. Default auto-resize
  final double height;

  /// border-radius of tag
  final BorderRadius borderRadius;

  /// box-shadow of tag
  final List<BoxShadow> boxShadow;

  ///placeholder InputField
  final String placeholder;

  /// imposes the same width and the same number of columns for each row
  final bool symmetry;

  /// Sometimes you need to hide the textField
  final bool textFieldHidden;

  /// margin between the tag
  final EdgeInsets margin;

  /// padding of the tag
  final EdgeInsets padding;

  /// type of row alignment
  final MainAxisAlignment alignment;

  /// offset of width (default 0)
  final int offset;

  /// possibility to insert duplicates in the list
  final bool duplicate;

  /// TextStyle of the tag
  final TextStyle textStyle;

  /// TextStyle of the tag
  final bool autocorrect;

  /// font size, the height of the tag is proportional to the font size
  final double fontSize;

  /// custom Icon close
  final Icon icon;

  /// icon size of Icon close
  final double iconSize;

  /// padding of Icon close
  final EdgeInsets iconPadding;

  /// margin of Icon close
  final EdgeInsets iconMargin;

  /// color of Icon close
  final Color iconColor;

  /// background of Icon close
  final Color iconBackground;

  /// border-radius of Icon Size
  final BorderRadius iconBorderRadius;

  /// type of text overflow within the tag
  final TextOverflow textOverflow;

  /// lower-case text tag
  final bool lowerCase;

  /// background color tag
  final Color color;

  /// background color container
  final Color backgroundContainer;

  ///highlight-color. it is activated when trying to insert a duplicate
  final Color highlightColor;

  /// callback
  final OnDelete onDelete;

  /// callback
  final OnInsert onInsert;

  /// Suggestions List
  final List<String> suggestionsList;

  /// Popup Menu Items
  /// (String Tag)
  final PopupMenuBuilder popupMenuBuilder;

  /// On Selected Item
  /// (int id, String tag)
  final PopupMenuOnSelected popupMenuOnSelected;

  @override
  _InputTagsState createState() => _InputTagsState();
}

class _InputTagsState extends State<InputTags> {
  final GlobalKey _containerKey = GlobalKey();
  Orientation _orientation = Orientation.portrait;

  //duplicate highlighting
  int _check = -1;

  // Position for popupMenu
  Offset _tapPosition;

  List<String> _tags = [];

  double _width = 0;
  double _initFontSize = 14;
  double _initMargin = 3;
  double _initPadding = 10;
  double _initBorderRadius = 50;

  double _initPaddingIcon = 3;
  double _initMarginIcon = 8;

  @override
  void initState() {
    super.initState();

    _tags = widget.tags;
  }

  @override
  void dispose() {
    super.dispose();
  }

  //get the current width of the container
  void _getWidthContext() {
    WidgetsBinding.instance.addPostFrameCallback((_) {
      final keyContext = _containerKey.currentContext;
      if (keyContext != null) {
        final RenderBox box = keyContext.findRenderObject();
        final size = box.size;
        setState(() {
          _width = size.width;
        });
      }
    });
  }

  @override
  Widget build(BuildContext context) {
    // essential to avoid infinite loop of addPostFrameCallback
    if (MediaQuery.of(context).orientation != _orientation || _width == 0) {
      _getWidthContext();
      _orientation = MediaQuery.of(context).orientation;

      //Avoid jitter/jump -> fix issue #14
      return Container(key: _containerKey);
    }

    return Container(
      key: _containerKey,
      margin: EdgeInsets.symmetric(vertical: 5.0, horizontal: 0.0),
      color: widget.backgroundContainer ?? Colors.white,
      child: Column(
        children: _buildRows(),
      ),
    );
  }

  List<Widget> _buildRows() {
    List<Widget> rows = [];

    int columns = widget.columns;

    //margin of the tag
    double margin =
        (widget.margin != null) ? widget.margin.horizontal : _initMargin * 2;

    //padding of the tag
    double padding =
        widget.padding != null ? widget.padding.horizontal : _initPadding;
    padding = padding * (widget.fontSize.clamp(8, 20) / 14);

    int tagsLength = _tags.length + 1;
    int rowsLength = (tagsLength / columns).ceil();
    double fontSize = widget.fontSize ?? _initFontSize;

    //initial width tag
    double widthTag = 1;

    int start = 0;
    bool overflow;

    for (int i = 0; i < rowsLength; i++) {
      // Single Row
      List<Widget> row = [];

      //break row
      overflow = false;

      //width of the Tag
      double tmpWidth = 0;

      // final index of the current row
      int end = start + columns;

      // makes sure that 'end' does not exceed 'tagsLength'
      if (end >= tagsLength) end -= end - tagsLength;

      for (int j = start; j < end; j++) {
        if (!widget.symmetry && _tags.isNotEmpty) {
          String tag = _tags[j % (tagsLength - 1)];

          TextSize txtSize = TextSize(
              txt: tag, fontSize: fontSize * (tag.length < 2 ? 1.2 : 1));

          // Total width of the tag
          double txtWidth =
              txtSize.get().width + _widthIcon() + padding + margin;

          //sum of the width of each tag
          //widget.offset it is optional but in special cases allows you to improve the width of the tags
          tmpWidth += txtWidth + (widget.offset ?? 0);

          if (j > start && tmpWidth > _width) {
            start = j;
            overflow = true;
            rowsLength += 1;
            break;
          }

          widthTag = txtWidth;
        }

        row.add(_buildField(
            index: _tags.isNotEmpty ? j % (tagsLength - 1) : null,
            width: widthTag,
            last: (j + 1 == tagsLength)));
      }

      // Check overflow width
      if (!overflow) start = end;

      rows.add(Row(
        mainAxisAlignment: widget.alignment ??
            ((widget.symmetry)
                ? MainAxisAlignment.start
                : MainAxisAlignment.center),
        children: row,
      ));
    }

    return rows;
  }

  /// Build single Tag
  Widget _buildField({int index, double width, bool last = false}) {
    String tag = (index != null) ? _tags[index] : '';

    final RenderBox overlay = Overlay.of(context).context.findRenderObject();

    Widget textField = Visibility(
      visible: !widget.textFieldHidden,
      child: Flexible(
          flex: (widget.symmetry) ? 1 : width.ceil(),
          child: Container(
              margin: widget.margin ??
                  EdgeInsets.symmetric(horizontal: _initMargin, vertical: 6),
              width: 200,
              child: _textField())),
    );

    if (last || tag == null)
      return textField;
    else {
      Widget container = Container(
        margin: widget.margin ??
            EdgeInsets.symmetric(horizontal: _initMargin, vertical: 6),
        width: (widget.symmetry) ? _widthCalc() : width,
        height: widget.height ?? 29 * (widget.fontSize / 14),
        decoration: BoxDecoration(
          boxShadow: widget.boxShadow ??
              [
                BoxShadow(
                    color: Colors.black.withOpacity(0.2),
                    spreadRadius: 0.5,
                    blurRadius: 4,
                    offset: Offset(0, 1))
              ],
          borderRadius:
              widget.borderRadius ?? BorderRadius.circular(_initBorderRadius),
          color: _check == index
              ? ((widget.highlightColor ?? widget.color?.withRed(700)) ??
                  Colors.green.withRed(450))
              : (widget.color ?? Colors.green[400]),
        ),
        child: Row(
          mainAxisSize: MainAxisSize.min,
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: <Widget>[
            Flexible(
              fit: FlexFit.loose,
              flex: 1,
              child: Padding(
                padding: widget.padding ??
                    EdgeInsets.only(
                        left: (_initPadding) *
                            (widget.fontSize.clamp(8, 20) / 14)),
                child: Text(
                  tag,
                  overflow: widget.textOverflow ?? TextOverflow.fade,
                  softWrap: false,
                  style: _textStyle,
                ),
              ),
            ),
            FittedBox(
                fit: BoxFit.cover,
                child: GestureDetector(
                  child: Container(
                    padding:
                        widget.iconPadding ?? EdgeInsets.all(_initPaddingIcon),
                    margin: widget.iconMargin ??
                        EdgeInsets.only(
                            right: _initMarginIcon *
                                (widget.fontSize != null
                                    ? (widget.fontSize.clamp(8, 22) / 26)
                                    : 1)),
                    decoration: BoxDecoration(
                      color: widget.iconBackground ?? Colors.transparent,
                      borderRadius: widget.iconBorderRadius ??
                          BorderRadius.circular(_initBorderRadius),
                    ),
                    child: Icon(widget.icon ?? Icons.clear,
                        color: widget.iconColor ?? Colors.white,
                        size: ((widget.fontSize != null)
                            ? 15 +
                                (widget.fontSize.clamp(6, 25).toDouble() - 18)
                            : 14)),
                  ),
                  onTap: () {
                    _check = -1;
                    if (widget.onDelete != null) widget.onDelete(tag);
                    setState(() {
                      _tags.remove(tag);
                    });
                  },
                ))
          ],
        ),
      );

      return Flexible(
          flex: (widget.symmetry) ? 1 : width.round(),
          child: GestureDetector(
            onTapDown: (details) {
              _tapPosition = details.globalPosition;
            },
            onLongPress: () {
              showMenu(
                      semanticLabel: tag,
                      items: widget.popupMenuBuilder(tag) ?? [],
                      context: context,
                      position: RelativeRect.fromRect(
                          _tapPosition & Size(40, 40),
                          Offset.zero &
                              overlay
                                  .size) // & RelativeRect.fromLTRB(65.0, 40.0, 0.0, 0.0),
                      )
                  .then((value) {
                if (widget.popupMenuOnSelected != null)
                  widget.popupMenuOnSelected(value, tag);
              });
            },
            child: widget.popupMenuBuilder == null
                ? Tooltip(
                    message: tag,
                    child: container,
                  )
                : container,
          ));
    }
  }

  /// TextFiled or CustomTextField
  Widget _textField() {
    return InputSuggestions(
      fontSize: widget.fontSize ?? null,
      suggestions: widget.suggestionsList ?? null,
      autofocus: widget.autofocus ?? true,
      keyboardType: widget.keyboardType ?? null,
      maxLength: widget.maxLength ?? null,
      lowerCase: widget.lowerCase ?? null,
      autocorrect: widget.autocorrect ?? false,
      onSubmitted: (str) {
        setState(() {
          _check = -1;
          if (_tags.contains(str) && !widget.duplicate)
            _check = _tags.indexWhere((st) => st == str);
          else {
            if (widget.onInsert != null) widget.onInsert(str);
            // fix issue #13
            if (!_tags.contains(str)) _tags.add(str);
          }
        });
      },
      style: TextStyle(color: Colors.black),
      inputDecoration: widget.inputDecoration ??
          InputDecoration(
              disabledBorder: InputBorder.none,
              errorBorder: InputBorder.none,
              contentPadding: EdgeInsets.symmetric(
                  vertical: 7 + (widget.fontSize.clamp(10, 24).toDouble() - 14),
                  horizontal:
                      10 + (widget.fontSize.clamp(10, 24).toDouble() - 14)),
              hintText: widget.placeholder ?? 'Add a tag',
              focusedBorder: UnderlineInputBorder(
                borderRadius: BorderRadius.circular(
                  _initBorderRadius,
                ),
                borderSide: BorderSide(
                  color: widget.color ?? Colors.green[400],
                ),
              ),
              enabledBorder: UnderlineInputBorder(
                borderRadius: BorderRadius.circular(
                  _initBorderRadius,
                ),
                borderSide: BorderSide(color: Colors.green.withOpacity(0.5)),
              ),
              border: UnderlineInputBorder(
                borderRadius: BorderRadius.circular(
                  _initBorderRadius,
                ),
                borderSide: BorderSide(color: Colors.green.withOpacity(0.5)),
              )),
    );
  }

  ///TextStyle
  TextStyle get _textStyle {
    if (widget.textStyle != null)
      return widget.textStyle.apply(
        color: widget.textStyle.color ?? Colors.white,
      );

    return TextStyle(
        fontSize: widget.fontSize ?? null,
        color: Colors.white,
        fontWeight: FontWeight.normal);
  }

  ///Total width of the close icon
  double _widthIcon() {
    double margin = widget.iconMargin != null
        ? widget.iconMargin.horizontal
        : _initMarginIcon;
    double padding = widget.iconPadding != null
        ? widget.iconPadding.horizontal / 2
        : _initPaddingIcon;
    double size = (widget.fontSize != null)
        ? 15 + (widget.fontSize.clamp(6, 25).toDouble() - 18)
        : 14;

    return margin + padding + size;
  }

  ///Container width divided by the number of columns when symmetry is active
  double _widthCalc() {
    int columns = widget.columns;

    int margin = (widget.margin != null)
        ? widget.margin.horizontal.round()
        : _initMargin.round() * 2;

    int subtraction = columns * (margin);
    double width = (_width > 1) ? (_width - subtraction) / columns : _width;

    return width;
  }
}
